v0.1.13: const incorrectly added to pointer parameter that is modified

## Version
v0.1.13

## Problem
When generating pointer parameters, `const` is added even when the pointed-to value is modified by the function.

## Reproduction

crc32.cnx:
```cnx
u32 crcFloat(u32 crc, f32 value) {
    string<4> buf <- "";
    buf[0, 4] <- value;
    crc <- this.crcByte(crc, buf[0]);  // crc is modified!
    crc <- this.crcByte(crc, buf[1]);
    crc <- this.crcByte(crc, buf[2]);
    crc <- this.crcByte(crc, buf[3]);
    return crc;
}
```

Generated crc32.cpp (wrong - has const):
```cpp
static uint32_t crc32_crcFloat(const uint32_t* crc, float value) {
    // ...
    (*crc) = crc32_crcByte((*crc), buf[0]);  // Error: assignment of read-only location
}
```

## Error
```
src/crc32.cpp:32:12: error: assignment of read-only location '* crc'
```

## Expected
Since `crc` is modified via `crc <- ...`, it should not be `const`:
```cpp
static uint32_t crc32_crcFloat(uint32_t* crc, float value) {
```

## Workaround
Manually remove `const` from the generated code.

--- a/src/codegen/CodeGenerator.ts
+++ b/src/codegen/CodeGenerator.ts
@@ -1001,13 +1001,6 @@ export default class CodeGenerator implements IOrchestrator {
   }
 
   /**
-   * Issue #281: Clear modified parameters tracking for a new function.
-   */
-  clearModifiedParameters(): void {
-    this.context.modifiedParameters.clear();
-  }
-
-  /**
    * Issue #268: Check if a callee function's parameter at given index is modified.
    * Returns true if the callee modifies that parameter (should not have const).
    */
@@ -4226,12 +4219,8 @@ export default class CodeGenerator implements IOrchestrator {
         // Track parameters for ADR-006 pointer semantics
         this._setParameters(funcDecl.parameterList() ?? null);
 
-        // Issue #281: Clear modified parameters tracking for this function
-        this.context.modifiedParameters.clear();
-
-        // ADR-016: Clear local variables and mark that we're in a function body
-        this.context.localVariables.clear();
-        this.context.inFunctionBody = true;
+        // ADR-016: Enter function body context (also clears modifiedParameters for Issue #281)
+        this.enterFunctionBody();
 
         // Issue #281: Generate body FIRST to track parameter modifications,
         // then generate parameter list using that tracking info
@@ -4245,9 +4234,8 @@ export default class CodeGenerator implements IOrchestrator {
           ? this._generateParameterList(funcDecl.parameterList()!)
           : "void";
 
-        // ADR-016: Clear local variables and mark that we're no longer in a function body
-        this.context.inFunctionBody = false;
-        this.context.localVariables.clear();
+        // ADR-016: Exit function body context
+        this.exitFunctionBody();
         this.context.currentFunctionName = null; // Issue #269: Clear function name
         this._clearParameters();
 

--- a/src/codegen/generators/IOrchestrator.ts
+++ b/src/codegen/generators/IOrchestrator.ts
@@ -304,12 +304,6 @@ interface IOrchestrator {
   markParameterModified(paramName: string): void;
 
   /**
-   * Issue #281: Clear modified parameters tracking for a new function.
-   * Call this before generating a function body to track fresh modifications.
-   */
-  clearModifiedParameters(): void;
-
-  /**
    * Issue #268: Check if a callee function's parameter at given index is modified.
    * Returns true if the callee modifies that parameter (should not have const).
    * Returns false if unmodified or unknown (callee not yet processed).

--- a/src/codegen/generators/declarationGenerators/ScopeGenerator.ts
+++ b/src/codegen/generators/declarationGenerators/ScopeGenerator.ts
@@ -130,10 +130,7 @@ const generateScope: TGeneratorFn<Parser.ScopeDeclarationContext> = (
       // Track parameters for ADR-006 pointer semantics
       orchestrator.setParameters(funcDecl.parameterList() ?? null);
 
-      // Issue #281: Clear modified parameters tracking for this function
-      orchestrator.clearModifiedParameters();
-
-      // ADR-016: Enter function body context
+      // ADR-016: Enter function body context (also clears modifiedParameters for Issue #281)
       orchestrator.enterFunctionBody();
 
       // Issue #281: Generate body FIRST to track parameter modifications,

--- a/tests/pass-by-value/scope-modified-param.expected.c
+++ b/tests/pass-by-value/scope-modified-param.expected.c
@@ -7,7 +7,9 @@
 
 // test-execution
 // Issue #281: const incorrectly added to pointer parameter that is modified
-// Tests that scope functions correctly omit const for modified pointer parameters
+// Tests that scope functions correctly handle parameter passing:
+// - Modified scalars: passed as pointer (no const) for mutation
+// - Unmodified scalars: passed by value (optimization over const pointer)
 /* Scope: TestScope */
 
 void TestScope_increment(uint32_t* x) {
@@ -41,6 +43,15 @@ void TestScope_switchIncrement(uint32_t* x, uint32_t mode) {
     }
 }
 
+uint32_t TestScope_readOnly(uint32_t x) {
+    return x * 2;
+}
+
+uint32_t TestScope_mixedParams(uint32_t* counter, uint32_t multiplier) {
+    (*counter) = (*counter) + 1;
+    return (*counter) * multiplier;
+}
+
 int main(void) {
     uint32_t counter = 10;
     TestScope_increment(&counter);
@@ -61,5 +72,15 @@ int main(void) {
     switchVal = 100;
     TestScope_switchIncrement(&switchVal, 99);
     if (switchVal != 101) return 7;
+    uint32_t readVal = 25;
+    uint32_t readResult = TestScope_readOnly(readVal);
+    if (readVal != 25) return 8;
+    if (readResult != 50) return 9;
+    uint32_t mixedCounter = 10;
+    uint32_t mixedMult = 3;
+    uint32_t mixedResult = TestScope_mixedParams(&mixedCounter, mixedMult);
+    if (mixedCounter != 11) return 10;
+    if (mixedMult != 3) return 11;
+    if (mixedResult != 33) return 12;
     return 0;
 }

--- a/tests/pass-by-value/scope-modified-param.test.cnx
+++ b/tests/pass-by-value/scope-modified-param.test.cnx
@@ -1,6 +1,8 @@
 // test-execution
 // Issue #281: const incorrectly added to pointer parameter that is modified
-// Tests that scope functions correctly omit const for modified pointer parameters
+// Tests that scope functions correctly handle parameter passing:
+// - Modified scalars: passed as pointer (no const) for mutation
+// - Unmodified scalars: passed by value (optimization over const pointer)
 
 scope TestScope {
     public void increment(u32 x) {
@@ -31,6 +33,17 @@ scope TestScope {
                 x <- x + 1;
             }
         }
+    }
+
+    // Negative test: unmodified params should still get const
+    public u32 readOnly(u32 x) {
+        return x * 2;  // x is NOT modified, should have const
+    }
+
+    // Mixed: first param modified, second param read-only
+    public u32 mixedParams(u32 counter, u32 multiplier) {
+        counter <- counter + 1;  // modified, no const
+        return counter * multiplier;  // multiplier only read
     }
 }
 
@@ -65,6 +78,20 @@ u32 main() {
     switchVal <- 100;
     TestScope.switchIncrement(switchVal, 99);
     if (switchVal != 101) return 7;
+
+    // Negative test: readOnly should NOT modify input (const param)
+    u32 readVal <- 25;
+    u32 readResult <- TestScope.readOnly(readVal);
+    if (readVal != 25) return 8;  // Should be unchanged
+    if (readResult != 50) return 9;  // Should be 25 * 2
+
+    // Mixed params: counter modified, multiplier const
+    u32 mixedCounter <- 10;
+    u32 mixedMult <- 3;
+    u32 mixedResult <- TestScope.mixedParams(mixedCounter, mixedMult);
+    if (mixedCounter != 11) return 10;  // counter was modified
+    if (mixedMult != 3) return 11;  // multiplier unchanged
+    if (mixedResult != 33) return 12;  // (10+1) * 3 = 33
 
     return 0;
 }

